### PR TITLE
[InstCombine] Propagate flags when folding consecutative shifts

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstCombine] Add tests for propagating flags when folding consecutative shifts; NFC**
- **[InstCombine] Propagate flags when folding consecutative shifts**

When we fold `(shift (shift C0, x), C1)` we can propagate flags that
are common to both shifts.

Proofs: https://alive2.llvm.org/ce/z/LkEzXD

### CODE REVIEW

**Review by nikic — APPROVED** [review]
LGTM

(It looks like your alive2 link also has the proofs for flag-preservation for a related case that's not part of this PR.)

**Comment by goldsteinn** [comment]
> LGTM
> 
> (It looks like your alive2 link also has the proofs for flag-preservation for a related case that's not part of this PR.)

Bah sorry for the confusion. Realized that case is already handled after writing the proofs.

File: llvm/test/Transforms/InstCombine/shift.ll [hunk 1]
```diff
@@ -2240,4 +2240,84 @@ define i129 @shift_zext_not_nneg(i8 %arg) {
...
+  %r = lshr i8 %sh, 1
+  ret i8 %r
+}
+
+define i8 @src_ashr_exact(i8 %x) {
+; CHECK-LABEL: @src_ashr_exact(
+; CHECK-NEXT:    [[R:%.*]] = lshr exact i8 8, [[X:%.*]]
+; CHECK-NEXT:    ret i8 [[R]]
+;
+  %sh = ashr exact i8 32, %x
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/shift.ll:2308]: Can you replace LHS with a negative number?

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Pull request closed without merging**